streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -250,127 +237,132 @@ abstract class ZSink[-R, +E, -I, +Z] private (
    * and emits the sink's results as outputs. The sink will be restarted when
    * it ends.
    */
-  def toTransducer: ZTransducer[R, E, I, Z] =
+  def toTransducer(implicit ev: L <:< I): ZTransducer[R, E, I, Z] =
     ZTransducer {
       ZSink.Push.restartable(push).map {
         case (push, restart) =>
-          (input: Option[Chunk[I]]) =>
+          def go(input: Option[Chunk[I]]): ZIO[R, E, Chunk[Z]] =
             push(input).foldM(
               {
-                case Left(e)  => ZIO.fail(e)
-                case Right(z) => restart.as(Chunk.single(z))
+                case (Left(e), _) => ZIO.fail(e)
```

Comment:
This is a weak spot. We have no contract that says that a transducer cannot be called after failing (and we shouldn't have such a contract). That means we're dropping leftovers inputs after failing.

How about placing these leftovers in a `Ref` that will be read on every push?

Reply:
I don't disagree with you, although I can't come up with a testcase so far.
Seems that once you called `stream.transduce(t)` you can't see beyond the first error.